node/network/statement-distribution/src/vstaging/grid.rs

```diff
@@ -0,0 +1,2247 @@
+	/// threshold.
+	Insufficient,
+	/// The manifest is malformed.
+	Malformed,
+	/// The manifest was not allowed to be sent.
+	Disallowed,
+}
+
+/// The knowledge we are aware of counterparties having of manifests.
+#[derive(Default)]
+struct ReceivedManifests {
+	received: HashMap<CandidateHash, ManifestSummary>,
+	// group -> seconded counts.
+	seconded_counts: HashMap<GroupIndex, Vec<usize>>,
+}
+
+impl ReceivedManifests {
+	fn candidate_statement_filter(
+		&self,
+		candidate_hash: &CandidateHash,
+	) -> Option<StatementFilter> {
+		self.received.get(candidate_hash).map(|m| m.statement_knowledge.clone())
+	}
+
+	/// Attempt to import a received manifest from a counterparty.
+	///
+	/// This will reject manifests which are either duplicate, conflicting,
+	/// or imply an irrational amount of `Seconded` statements.
+	///
+	/// This assumes that the manifest has already been checked for
+	/// validity - i.e. that the bitvecs match the claimed group in size
+	/// and that that the manifest includes at least one `Seconded`
+	/// attestation and includes enough attestations for the candidate
+	/// to be backed.
+	///
+	/// This also should only be invoked when we are intended to track
+	/// the knowledge of this peer as determined by the [`SessionTopology`].
+	fn import_received(
+		&mut self,
+		group_size: usize,
+		seconding_limit: usize,
+		candidate_hash: CandidateHash,
+		manifest_summary: ManifestSummary,
+	) -> Result<(), ManifestImportError> {
+		match self.received.entry(candidate_hash) {
+			Entry::Occupied(mut e) => {
+				// occupied entry.
+
+				// filter out clearly conflicting data.
+				{
+					let prev = e.get();
+					if prev.claimed_group_index != manifest_summary.claimed_group_index {
+						return Err(ManifestImportError::Conflicting)
+					}
+
+					if prev.claimed_parent_hash != manifest_summary.claimed_parent_hash {
+						return Err(ManifestImportError::Conflicting)
+					}
+
+					if !manifest_summary
+						.statement_knowledge
+						.seconded_in_group
+						.contains(&prev.statement_knowledge.seconded_in_group)
+					{
+						return Err(ManifestImportError::Conflicting)
+					}
+
+					if !manifest_summary
+						.statement_knowledge
+						.validated_in_group
+						.contains(&prev.statement_knowledge.validated_in_group)
+					{
+						return Err(ManifestImportError::Conflicting)
+					}
+
+					let mut fresh_seconded =
+						manifest_summary.statement_knowledge.seconded_in_group.clone();
+					fresh_seconded |= &prev.statement_knowledge.seconded_in_group;
+
+					let within_limits = updating_ensure_within_seconding_limit(
+						&mut self.seconded_counts,
+						manifest_summary.claimed_group_index,
+						group_size,
+						seconding_limit,
+						&fresh_seconded,
+					);
+
+					if !within_limits {
+						return Err(ManifestImportError::Overflow)
+					}
+				}
+
+				// All checks passed. Overwrite: guaranteed to be
+				// superset.
+				*e.get_mut() = manifest_summary;
```

Comment:
What's the case of overwriting a manifest?

Reply:
not currently implemented, but it's allowed in this helper to send a manifest multiple times, as long as the new manifest contains a superset of the statements of the first (checked above)